### ISSUE TITLE
Black Screen Bug

### ISSUE THREAD

**Issue by biblicabeebli** [onnela-lab/beiwe-ios#40]
The black screen bug has existed for years. We don't know when it was introduced, and we can't bisect because of swift version requirements in newer versions of xcode/ios. We also cannot always reproduce it from user reported actions.

The behavior is: with the app is open in the background for some time, the user reopens the app is presented with an entirely black screen. Thus, The Black Screen Bug.

This kind of bug is _something_ to do with a uiview (storyboard?) transition - or at least that's what the internet says it is - and our behavior matches posted issues I've found, but the listed fixes never apply.

It happens a lot around tapping push notifications, but after exhaustive testing I can state that this is just happenstance. The ui lockout occurs after push notification logic runs. People mostly open the app by tapping on push notifications.

Our codebase has very few storyboard transitions. This bug exclusively occurs after registration, which limits it to a handful of interactions at app (re)launch. Most of that code is contained within the `transitionToLoadedAppState` function in AppDelegate, which contains calls to our `changeRootViewController` and `changeRootViewControllerWithIdentifier` that do the raw work. `initializeUI` is only called when the app starts and it sets it to `launchScreen`, and isn't called at _re_launch.

It looks like we also have a `displayCurrentMainView` that is never called. 🧐

After ruling out push notification code itself I restarted testing with the goal of triggering the automatic logout timer by setting it to 30 seconds for my debugging study.  This _very quickly_ triggered the bug - hooray! Something-anything to look into.

---

The contents of `transitionToLoadedAppState` is not complex; its purpose is to centralize the last stage of loading the app, which needs to be triggered from several places.  It sets some variables, checks whether a study is loaded, and calls one of the `changeRootViewController` functions based on app state.  The black screen bug happens only when the logout timer has expired.  There is absolutely nothing special about the code surrounding that.

This is the function, with irrelevant sections removed, updated comments for clarity:

```swift
func transitionToLoadedAppState() {
    if let currentStudy = StudyManager.sharedInstance.currentStudy {
        if currentStudy.participantConsented {
            StudyManager.sharedInstance.startStudyDataServices()  // complex, needs to be called before most of the app works
        }
        if !self.isLoggedIn {
            self.changeRootViewControllerWithIdentifier("login")  // Load up the log in view
        } else {
            if currentStudy.participantConsented {
                self.changeRootViewControllerWithIdentifier("mainView")  // load main interface,
            } else {
                self.changeRootViewController(ConsentManager().consentViewController)
            }
        }
    } else {
        // no study go to onboarding view
        self.changeRootViewController(OnboardingManager().onboardingViewController)
    }
}
```

So the bug always(?) only(?) happens when we hit `self.changeRootViewControllerWithIdentifier("login")`?  Before my testing this value was set to 10 minutes, and I think I could hit it with less then 10 minutes of backgrounded app passing.

---

***some debugging time later***

We always(?) get this message in the app log when the black screen appears (there's no file location given, why would we need that that seems silly and helpful):
`This method can cause UI unresponsiveness if invoked on the main thread. Instead, consider waiting for the '-locationManagerDidChangeAuthorization:' callback and checking 'authorizationStatus' first.`

So there's some call to location manager code on the main thread ... but changing the uiview, which `changeRootViewController` is the active code on the main thread. In fact it _has to use the main thread, the app crashes if it isn't._ **So why is there unrelated code running on the main thread?**

Oh gawd its a race condition with `async` code.  Of course it is.
It _shouldn't_ be. The location manager code I've fiddled with causes like a half second stall, not permanent app lockup, but it's screwing something up.

Wrapping this specific line of code in an main-thread-async call fixes it.

```swift
DispatchQueue.main.async {
    self.changeRootViewControllerWithIdentifier("login")
}
```

Well, that's stupid.
This isn't the end of the bug, I should probably wrap the contents of changeRootViewController in a main thread async call rather than this single instance of it, but we will start by testing this internally.

**Comment by biblicabeebli** [comment]
Shifted location of the fix to be inside `changeRootViewController` to cover any future cases of this issue (this is a hope, not a certainty).

Tested transition animations and found that the login check mark animation was getting cut off, fixed that issue.

Most recent work is on branch 2.4.12, just got testflight notification that its up.

**Comment by biblicabeebli** [comment]
These changes appear to be completely benign, we are now pending review on the App Store.

**Comment by biblicabeebli** [comment]
And it's on the app store as of some time over the weekend.

**Comment by biblicabeebli** [comment]
Based on review of my logs of old black screen bugs that I kept around it looks like the authorizationStatus message I noticed and assumed was the proximate cause of the race condition was NOT always present.

If true (I definitely purged some lines from those logs) then we still have an unknown proximate cause / source of the thing I'm going to continue to call a race condition.

I can't do anything to reduce ambiguity as THIS black screen bug is fixed.  I'm going to close this issue, if anything comes up it will have to be reexamined.  I've added a note as a comment in the codebase next to the existing fix.
